timlee/networking-with-a-raspberry/README.md

```diff
@@ -0,0 +1,280 @@
+
+- `sudo nano /etc/nginx/sites-available/default` and edit the original rule for port 80 to remove the entire `location` subsection and replace it with the single line `return 301 https://$host$request_uri;`
+- `sudo /etc/init.d/nginx restart` to restart the server
+
+Now try reaching the site over HTTP (`http://badger.facts`) and watch as it magically gets upgraded to HTTPS. Lovely!
+
+## Step 4: Reverse proxy
+
+A reverse proxy can improve performance, increase flexibility and boost security - let's try it out.
+
+### What is a reverse proxy?
+
+Reverse proxies in particular always frightened me - I thought that no one really understood them and people just used phrases like 'reverse proxy' to sound smart and scare junior devs. It turns out, they're actually really simple.
+
+![A forward vs a reverse proxy](./proxies.png)
+
+A standard (forward) proxy server sits in front of a network and relays requests from the machines on the network out to the internet, returning the responses to each machine.
+
+A reverse proxy is the opposite - it accepts incoming client requests from the internet and sends them to the correct machine on the network, returning the machine's response to the client. It's pretty much that simple.
+
+The reverse proxy becomes the public interface for a site - the address of the proxy is the one that DNS servers will hand out to clients.
+
+There are many benefits that this extra layer can bring:
+
+- caching and SSL termination for decreased workloads on the destination servers
+- compression for reduced payloads and response times
+- decoupling of the publically accessible sites from the actual server configuration for increased flexibility of the infrastructure without public impact
+- increased security from reduced leakage of server configuration outside of the network
+
+We can configure this for the Raspberry Pi quite easily - Nginx loves being a reverse proxy!
+
+### Configuring the proxy
+
+At the moment, we have a single site configured, listening on ports 80 and 443 to requests for all domains.
+
+Let's update that to create a dedicated configuration for our site.
+
+- Make a copy of the `default` file and name it according to your website, eg `badger.facts.conf`
+- Update the configuration for the `server_name` (which is used to match the domain on incoming requests) from `_` (a wildcard) to that of your site (i.e. `badger.facts`). Do this for both the port 80 and port 443 server sections
+- Replace the contents of the `location` section for the 443 server to `proxy_pass http://badger.facts:8888;`. This will tell Nginx that this server will act as a reverse proxy, passing the request transparently through to a server listening on port 8888
+- Create a new `server` section at the bottom of the file, which will act as the new internal web server that will respond to requests that are passed through from the reverse proxy. It should listen on port 8888 and respond with the index page
+
+```
```

Comment:
add nginx markdown flavour

```nginx
server {
  listen 443;
  server_name badger.facts;
  ssl on;
  ssl_certificate /etc/ssl/badger-facts-bundle.crt;
  ssl_certificate_key /etc/ssl/badger-facts.key;
  location / {
    proxy_pass http://badger.facts:8888;
  }
}
server {
  listen 8888;
  server_name badger.facts;
  location / {
    root /var/www/html;
    index index.nginx-debian.html;
  }
}
```

Reply:
Loved how simple and plain was explained. I added some comments about style and some suggestions to add more references